File: hackerrank/problem solving/algorithms/2 - implementation/1 - grading students/with morfo library/grading_students.go
// https://www.hackerrank.com/challenges/grading/problem?isFullScreen=true

package main

import "morfo/morfoio"

func main() {
    n := morfoio.Read[int]()
    grades := morfoio.ReadN[int](n)
    morfoio.Println(gradingStudents(grades))
}

// n: length of grades
// T: O(n)
// S: O(1) extra space
func gradingStudents(grades []int) []int {
    const minGrade = 38

    for i, grade := range grades {
        if grade >= minGrade {
            if nextMultiple5 := (grade/5 + 1) * 5; nextMultiple5-grade < 3 {
                grades[i] = nextMultiple5
            }
        }
    }

    return grades
}

File: hackerrank/problem solving/algorithms/2 - implementation/1 - grading students/with morfo library/grading_students.ts
// https://www.hackerrank.com/challenges/grading/problem?isFullScreen=true

import { idiv, println, read } from 'morfo';

const n: number = read(Number);
const grades: number[] = read(n, Number);
println(gradingStudents(grades));

// n: length of grades
// T: O(n)
// S: O(1) extra space
function gradingStudents(grades: number[]): number[] {
  for (let i = 0, minGrade = 38; i < grades.length; i++) {
    if (grades[i] < minGrade) continue;
    const nextMultiple5 = (idiv(grades[i], 5) + 1) * 5;
    if (nextMultiple5 - grades[i] < 3) grades[i] = nextMultiple5;
  }
  return grades;
}
